compiler/src/dotty/tools/dotc/typer/Checking.scala

```diff
@@ -679,20 +679,28 @@ trait Checking {
     tree.tpe.widenTermRefExpr match {
       case tp: ConstantType if exprPurity(tree) >= purityLevel => // ok
       case tp =>
+        def isCaseClassApply(sym: Symbol): Boolean =
+          sym.name == nme.apply && sym.owner.is(Module) && sym.owner.companionClass.is(Case)
+        def isCaseClassNew(sym: Symbol): Boolean =
+          sym.isPrimaryConstructor && sym.owner.is(Case) && sym.owner.isStatic
```

Comment:
`.owner.isStatic` is tested both here and inside the allow condition

Reply:
I will remove one of them